File: HomeWorkProjects/homework3/src/main/java/com/gmail/petrusevich/volha/homework3/contacts/activity/MainActivity.java
package com.gmail.petrusevich.volha.homework3.contacts.activity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.recyclerview.widget.RecyclerView;

import com.gmail.petrusevich.volha.homework3.R;
import com.gmail.petrusevich.volha.homework3.contacts.adapter.ContactListAdapter;
import com.gmail.petrusevich.volha.homework3.contacts.editing.EditListContacts;
import com.gmail.petrusevich.volha.homework3.contacts.listcontacts.Contacts;
import com.gmail.petrusevich.volha.homework3.contacts.sorting.SearchContact;
import com.gmail.petrusevich.volha.homework3.contacts.sorting.SortedContact;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements View.OnClickListener, ContactListAdapter.OnContactListener {

    private RecyclerView recyclerView;
    private List<Contacts> contacts = new ArrayList<>();
    private FloatingActionButton button;
    private int position;
    private TextView noContact;
    private SortedContact sortedContact = new SortedContact(contacts);
    private SearchView searchView;
    private SearchContact searchContact = new SearchContact();
    private ContactListAdapter contactListAdapter;
    private EditListContacts editListContacts = new EditListContacts();
    private final String SAVE_KEY = "saveListContacts";
    private final String SAVE_SEARCH_KEY = "saveSearchContacts";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findView();
        button.setOnClickListener(this);
        sortedContact.sortContact(contacts);
        contactListAdapter = new ContactListAdapter(contacts, this);
        recyclerView.setAdapter(contactListAdapter);
        emptyList();
        searchContact.searchContact(searchView, contactListAdapter);
    }

    private int getPosition() {
        return position;
    }

    private void setPosition(int position) {
        this.position = position;
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.add_button) {
            startActivityForResult(AddContactActivity.newIntent(this), 1);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data == null) {
            return;
        } else if (requestCode == 1) {
            editListContacts.addContact(data, contacts, contactListAdapter);
            emptyList();
        } else if (requestCode == 2) {
            editListContacts.editContact(data, contacts, contactListAdapter, getPosition());
            editListContacts.removeContact(data, contacts, contactListAdapter, position);
            emptyList();
        }
    }

    @Override
    public void onContactClick(int position) {
        setPosition(position);
        startActivityForResult(EditContactActivity.newIntent(this, contacts.get(position)), 2);
    }

    private void emptyList() {
        if (contacts.isEmpty()) {
            recyclerView.setVisibility(View.GONE);
            noContact.setVisibility(View.VISIBLE);
        } else {
            recyclerView.setVisibility(View.VISIBLE);
            noContact.setVisibility(View.GONE);
        }
    }

    private void findView() {
        recyclerView = findViewById(R.id.list_contact);
        button = findViewById(R.id.add_button);
        noContact = findViewById(R.id.noContactText);
        searchView = findViewById(R.id.search_contact);
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(SAVE_SEARCH_KEY, searchView.getQuery().toString());
        outState.putParcelableArrayList(SAVE_KEY, (ArrayList<Contacts>) contacts);
    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        contacts = savedInstanceState.getParcelableArrayList(SAVE_KEY);
        contactListAdapter = new ContactListAdapter(contacts, this);
        recyclerView.setAdapter(contactListAdapter);
        searchContact.searchContact(searchView, contactListAdapter);
        emptyList();
        String t = savedInstanceState.getString(SAVE_SEARCH_KEY);
        searchView.setQuery(t, true);
    }

}

File: HomeWorkProjects/homework5/src/main/java/com/gmail/petrusevich/volha/homework5/MainActivity.kt
package com.gmail.petrusevich.volha.homework5

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import com.gmail.petrusevich.volha.homework5.adapter.ContactListAdapter
import com.gmail.petrusevich.volha.homework5.database.controller.ContactViewModel
import com.gmail.petrusevich.volha.homework5.adapter.OnContactListener
import com.gmail.petrusevich.volha.homework5.database.ContactDatabase
import com.gmail.petrusevich.volha.homework5.database.Contacts
import com.gmail.petrusevich.volha.homework5.database.controller.SearchContact
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), View.OnClickListener, OnContactListener {

    private val listContacts: List<Contacts> = mutableListOf()
    private lateinit var contactDatabase: ContactDatabase
    private lateinit var viewModel: ContactViewModel
    private val searchContact: SearchContact = SearchContact()
    private lateinit var recyclerView: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        viewAddButton.setOnClickListener(this)
        contactDatabase = ContactDatabase.getInstance(this)!!
        recyclerView = findViewById<RecyclerView>(R.id.viewListContact)
        val contactListAdapter = ContactListAdapter(listContacts, this)
        recyclerView.adapter = contactListAdapter
        viewModel = ViewModelProvider(this).get(ContactViewModel::class.java)
        viewModel.listContacts?.observe(this, Observer {
            it.let {
                contactListAdapter.updateListContact(it)
                if (it.isEmpty()) {
                    recyclerView.visibility = View.GONE
                    viewNoContactText.visibility = View.VISIBLE
                } else {
                    recyclerView.visibility = View.VISIBLE
                    viewNoContactText.visibility = View.GONE
                }
            }
        })
        searchContact.searchContact(viewSearchContact, contactListAdapter, viewModel)
    }


    override fun onClick(view: View?) {
        if (view == viewAddButton) {
            startActivity(AddContactActivity.newIntent(this))
        }
    }

    override fun onContactClick(position: Int) {
        val contacts = viewModel.listContacts?.value
        startActivity(EditContactActivity.newIntent(this, contacts?.get(position)?.id))
    }

}
